src/reg.rs

```diff
@@ -0,0 +1,45 @@
+//! Register modifying support
+
+use core::sync::atomic::{compiler_fence, Ordering};
+
+/// Modify specific index of array-like register
+macro_rules! modify_at {
+    ($reg:expr, $bitwidth:expr, $index:expr, $value:expr) => {
+        $reg.modify(|r, w| {
+            let mask = !(u32::MAX >> (32 - $bitwidth) << ($bitwidth * $index));
+            let value = $value << ($bitwidth * $index);
+            w.bits(r.bits() & mask | value)
+        })
+    };
+}
+pub(crate) use modify_at;
+
+/// Modify specific index of array-like register atomically
+#[inline]
+pub(crate) unsafe fn atomic_modify_at(reg: *mut u32, bitwidth: u8, index: u8, value: u32) {
+    let mask_comp = u32::MAX >> (32 - bitwidth) << (bitwidth * index);
+    let value = value << (bitwidth * index);
+
+    // Perform volatile atomic `*reg = *reg & !mask_comp | value;`
+    let mut res: u32;
+    while {
+        compiler_fence(Ordering::SeqCst);
+        core::arch::asm!(
+            "ldrex {0}, [{1}]",
+            out(reg) res,
+            in(reg) reg,
+            options(readonly, preserves_flags, nostack),
+        );
+        compiler_fence(Ordering::SeqCst);
+        res = res & !mask_comp | value;
+        compiler_fence(Ordering::SeqCst);
+        core::arch::asm!(
+            "strex {0}, {0}, [{1}]",
+            inout(reg) res,
+            in(reg) reg,
+            options(preserves_flags, nostack),
+        );
+        compiler_fence(Ordering::SeqCst);
+        res != 0
```

Comment:
Without context, it is not clear what this comparison is fore. Maybe document under which conditions `res` is not `0` and why `0` is the goal.